langchain4j-weaviate/src/main/java/dev/langchain4j/store/embedding/weaviate/WeaviateEmbeddingStore.java

```diff
@@ -200,34 +229,62 @@ private List<String> addAll(List<String> ids, List<Embedding> embeddings, List<T
                     ? ids.get(i)
                     : avoidDups && embedded != null ? generateUUIDFrom(embedded.get(i).text()) : randomUUID();
             resIds.add(id);
-            objects.add(buildObject(id, embeddings.get(i), embedded != null ? embedded.get(i).text() : ""));
+            objects.add(buildObject(id, embeddings.get(i), embedded != null ? embedded.get(i) : null));     
         }
-
-        client
-                .batch()
-                .objectsBatcher()
+        client.batch().objectsBatcher()
                 .withObjects(objects.toArray(new WeaviateObject[0]))
                 .withConsistencyLevel(consistencyLevel)
                 .run();
-
         return resIds;
     }
 
-    private WeaviateObject buildObject(String id, Embedding embedding, String text) {
+    private WeaviateObject buildObject(String id, Embedding embedding, TextSegment segment) {
         Map<String, Object> props = new HashMap<>();
-        props.put(METADATA_TEXT_SEGMENT, text);
-
+        Map<String, Object> metadata = new HashMap<>();
+        if (segment != null) {
+            props.put(METADATA_TEXT_SEGMENT, segment.text());
+            if (segment.metadata() != null && !segment.metadata().asMap().isEmpty()) {
```

Comment:
```suggestion
            if (!segment.metadata().asMap().isEmpty()) {
```
metadata is always non-null